### PR TITLE
AG-4716 - added strictNullChecks and fixed sonarQube code smells

### PR DESCRIPTION
**Pull request by gportela85**
https://ag-grid.atlassian.net/browse/AG-4716

### CODE REVIEW

File: community-modules/core/src/ts/utils/generic.ts [hunk 1]
```diff
@@ -179,11 +185,11 @@ export function find<T>(collection: T[] | { [id: string]: T; }, predicate: strin
 
 export function values<T>(object: { [key: string]: T; } | Set<T> | Map<any, T>): T[] {
     if (object instanceof Set || object instanceof Map) {
-        const values: T[] = [];
+        const arr: T[] = [];
```
Comment (alowdon) [inline community-modules/core/src/ts/utils/generic.ts:188]: Why was this renamed from `values`?

File: community-modules/infinite-row-model/src/infiniteRowModel/infiniteRowModel.ts [hunk 2]
```diff
@@ -219,8 +216,8 @@ export class InfiniteRowModel extends BeanStub implements IInfiniteRowModel {
         this.infiniteCache = this.createBean(new InfiniteCache(this.cacheParams));
     }
 
-    private defaultIfInvalid(value: number, defaultValue: number): number {
-        return value > 0 ? value : defaultValue;
+    private defaultIfInvalid(value: number | undefined, defaultValue: number): number {
+        return value! > 0 ? value! : defaultValue;
```
Comment (alowdon) [inline community-modules/infinite-row-model/src/infiniteRowModel/infiniteRowModel.ts:220]: Does this need the `!` since the `> 0` will check for existence?

File: enterprise-modules/column-tool-panel/src/columnToolPanel/primaryColsListPanel.ts [hunk 3]
```diff
@@ -375,7 +376,9 @@ export class PrimaryColsListPanel extends Component {
         const passesFilter = (item: ColumnModelItem) => {
             if (!_.exists(this.filterText)) { return true; }
 
-            return item.getDisplayName() != null ? item.getDisplayName().toLowerCase().indexOf(this.filterText as string) >= 0 : true;
+            const displayName = item.getDisplayName();
+
+            return displayName != null ? displayName.toLowerCase().indexOf(this.filterText) !== -1 : true;
```
Comment (alowdon) [inline enterprise-modules/column-tool-panel/src/columnToolPanel/primaryColsListPanel.ts:381]: Could be `return displayName == null || displayName.toLowerCase().indexOf(this.filterText) !== -1`

File: enterprise-modules/excel-export/src/excelExport/files/ooxml/column.ts [hunk 4]
```diff
@@ -9,8 +9,8 @@ const columnFactory: ExcelOOXMLTemplate = {
         let excelWidth = 1;
         let customWidth = '0';
 
-        if (width > 1) {
-            excelWidth = getExcelCellWidth(width);
+        if (width! > 1) {
```
Comment (alowdon) [inline enterprise-modules/excel-export/src/excelExport/files/ooxml/column.ts:12]: If `> 1` doesn't satisfy SQ, then easy change to `width != null && width > 1`

File: enterprise-modules/excel-export/src/excelExport/files/ooxml/styles/border.ts [hunk 5]
```diff
@@ -79,7 +79,7 @@ export const convertLegacyBorder = (type: string, weight: number): string => {
 
     if (!type) { return 'thin'; }
     if (type === 'Continuous') { return namedWeight; }
-    if (namedWeight === 'medium' && mediumBorders.indexOf(mappedName) > 0) { return `medium${mappedName}`; }
+    if (namedWeight === 'medium' && mediumBorders.indexOf(mappedName) !== -1) { return `medium${mappedName}`; }
```
Comment (alowdon) [inline enterprise-modules/excel-export/src/excelExport/files/ooxml/styles/border.ts:82]: This is not the same logic - is this deliberate?

File: enterprise-modules/filter-tool-panel/src/filterToolPanel/toolPanelFilterComp.ts [hunk 6]
```diff
@@ -125,6 +125,7 @@ export class ToolPanelFilterComp extends Component {
             filterPromise.then(filter => {
                 this.underlyingFilter = filter;
 
+                if (!filter) { return ; }
```
Comment (alowdon) [inline enterprise-modules/filter-tool-panel/src/filterToolPanel/toolPanelFilterComp.ts:128]: Redundant space

File: enterprise-modules/row-grouping/src/rowGrouping/columnDropZones/dropZoneColumnComp.ts [hunk 7]
```diff
@@ -137,7 +136,7 @@ export class DropZoneColumnComp extends Component {
         if (this.valueColumn) {
             const aggFunc = this.column.getAggFunc();
             // if aggFunc is a string, we can use it, but if it's a function, then we swap with 'func'
-            const aggFuncString = (typeof aggFunc === 'string') ? aggFunc as string : 'agg';
+            const aggFuncString = (typeof aggFunc === 'string') ? aggFunc : 'agg';
```
Comment (alowdon) [inline enterprise-modules/row-grouping/src/rowGrouping/columnDropZones/dropZoneColumnComp.ts:139]: Superfluous brackets

File: community-modules/client-side-row-model/src/clientSideRowModel/clientSideRowModel.ts [hunk 8]
```diff
@@ -847,24 +847,24 @@ export class ClientSideRowModel extends BeanStub implements IClientSideRowModel
         return rowNodeTran;
     }
 
-    private createRowNodeOrder(): { [id: string]: number; } {
+    private createRowNodeOrder(): { [id: string]: number; } | undefined {
         const suppressSortOrder = this.gridOptionsWrapper.isSuppressMaintainUnsortedOrder();
         if (suppressSortOrder) { return; }
 
-        const orderMap: { [id: string]: number } = suppressSortOrder ? null : {};
+        const orderMap: { [id: string]: number } = {};
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/client-side-row-model/src/clientSideRowModel/clientSideRowModel.ts:854]: see line 852, there is no way supressSortOrder can be truthy

File: community-modules/client-side-row-model/src/clientSideRowModel/sortService.ts [hunk 9]
```diff
@@ -216,14 +211,12 @@ export class SortService extends BeanStub {
                     console.error('ag-Grid: groupHideOpenParents only works when specifying specific columns for colDef.showRowGroup');
                     return;
                 }
-                const displayingGroupKey: string = showRowGroup as string;
 
+                const displayingGroupKey = showRowGroup;
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/client-side-row-model/src/clientSideRowModel/sortService.ts:215]: line 208 is what guarantees this to be a string, so the variable is needed.

File: community-modules/core/src/ts/columnController/columnController.ts [hunk 10]
```diff
@@ -1023,10 +1030,12 @@ export class ColumnController extends BeanStub {
...
 
         columns.forEach(col => {
-            minWidthAccumulated += col.getMinWidth();
+            const minWidth = col.getMinWidth();
+            minWidthAccumulated += minWidth || 0;
 
-            if (col.getMaxWidth() > 0) {
-                maxWidthAccumulated += col.getMaxWidth();
+            const maxWidth = col.getMaxWidth();
+            if (exists(maxWidth) && maxWidth > 0) {
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/core/src/ts/columnController/columnController.ts:1037]: yes, or could do `maxWidth! > 0`. Both work.

File: community-modules/core/src/ts/columnController/columnController.ts [hunk 11]
```diff
@@ -2150,13 +2172,30 @@ export class ColumnController extends BeanStub {
         const getValue = (key1: string, key2?: string): { value1: any, value2: any; } => {
             const stateAny = stateItem as any;
             const defaultAny = defaultState as any;
-            if (stateAny && (stateAny[key1] !== undefined || stateAny[key2] !== undefined)) {
-                return { value1: stateAny[key1], value2: stateAny[key2] };
-            } else if (defaultAny && (defaultAny[key1] !== undefined || defaultAny[key2] !== undefined)) {
-                return { value1: defaultAny[key1], value2: defaultAny[key2] };
-            } else {
-                return { value1: undefined, value2: undefined };
+            const obj: { value1: any, value2: any } = { value1: undefined, value2: undefined };
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/core/src/ts/columnController/columnController.ts:2175]: If you use `{}` in strict mode there will be an error: "Type '{}' is missing the following properties from type '{ value1: any; value2: any; }': value1, value2"

File: community-modules/core/src/ts/components/componentUtil.ts [hunk 12]
```diff
@@ -180,25 +176,21 @@ export class ComponentUtil {
...
-            return value;
-        } else if (typeof value === 'string') {
-            return Number(value);
-        } else {
-            return undefined;
-        }
+    public static toNumber(value: any): number | undefined {
+        if (typeof value === 'number') { return value; }
+
+        if (typeof value === 'string') { return Number(value); }
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/core/src/ts/components/componentUtil.ts:193]: Yes, it shouts when we return undefined

File: community-modules/core/src/ts/filter/provided/providedFilter.ts [hunk 13]
```diff
@@ -162,18 +162,21 @@ export abstract class ProvidedFilter extends ManagedFocusComponent implements IF
...
             }
 
-            const button = loadTemplate(/* html */
+            const button = loadTemplate(
+                /* html */
                 `<button
                     type="button"
                     ref="${type}FilterButton"
-                    class="ag-standard-button ag-filter-apply-panel-button">${text}</button>`);
+                    class="ag-standard-button ag-filter-apply-panel-button">${text}>
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/core/src/ts/filter/provided/providedFilter.ts:177]: good catch

File: community-modules/core/src/ts/utils/browser.ts [hunk 14]
```diff
@@ -32,7 +32,7 @@ export function isBrowserSafari(): boolean {
         const anyWindow = window as any;
         const hasNotification = (p: any) => p && p.toString() === '[object SafariRemoteNotification]';
 
-        isSafari = Object.prototype.toString.call(anyWindow.HTMLElement).indexOf('Constructor') > 0
+        isSafari = Object.prototype.toString.call(anyWindow.HTMLElement).indexOf('Constructor') !== -1
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/core/src/ts/utils/browser.ts:35]: yes this was a mistake, but it's only used for older versions of Safari. New versions of Safari will use 
window.safari.pushNotification.toString() === '[object SafariRemoteNotification]'

File: community-modules/core/src/ts/utils/map.ts [hunk 15]
```diff
@@ -1,7 +1,7 @@
 export function keys<T>(map: Map<T, any>): T[] {
-    const keys: T[] = [];
+    const arr: T[] = [];
```
[the remark this answers is not in the document]
Comment (gportela85) [reply community-modules/core/src/ts/utils/map.ts:2]: SonarQube complains about having the same name within and outside of scope

File: community-modules/core/src/ts/utils/generic.ts (hunk 1 above)
Comment (gportela85) [reply community-modules/core/src/ts/utils/generic.ts:188]: same reason as keys within map.ts

File: community-modules/infinite-row-model/src/infiniteRowModel/infiniteRowModel.ts (hunk 2 above)
Comment (gportela85) [reply community-modules/infinite-row-model/src/infiniteRowModel/infiniteRowModel.ts:220]: That doesn't happen in strict mode

File: enterprise-modules/excel-export/src/excelExport/files/ooxml/styles/border.ts (hunk 5 above)
Comment (gportela85) [reply enterprise-modules/excel-export/src/excelExport/files/ooxml/styles/border.ts:82]: yes, it was a mistake before. This wouldn't detect `mediumDashed` borders
